terraform/main.tf

```diff
@@ -0,0 +1,42 @@
+resource "juju_application" "nms" {
+  name  = "nms"
+  model = var.model_name
+
+  charm {
+    name    = "sdcore-nms-k8s"
+    channel = var.channel
+  }
+  units = 1
+  trust = true
+}
+
+resource "juju_integration" "nms-traefik" {
+  model = var.model_name
+
+  application {
+    name     = juju_application.nms.name
+    endpoint = "ingress"
+  }
+
+  application {
+    name     = var.traefik_application_name
+    endpoint = "ingress"
+  }
+}
+
+resource "juju_integration" "nms-webui" {
+  model = var.model_name
+
+  application {
+    name     = juju_application.nms.name
+    endpoint = "sdcore-management"
+  }
+
+  application {
+    name     = var.webui_application_name
+    endpoint = "sdcore-management"
+  }
+}
```

Comment:
IIRC we shouldn't be defining integrations here. On the higher level, when we start importing modules, this will cause failures, because the other side will define the same integrations as well, right?

Reply:
The integrations are defined only in the requirer side, not in both side.  For example `nms` requires  `webui` and `traefik` but the reverse is not correct. In `webui` TF, it requires  only  `mongo`.  So, there will not any problem or conflict in the higher level.

Reply:
Yeah, I thought about it too, but I don't think it's the right way to go. Some of the "required" relations are in fact optional; like `ingress` in this case. We don't really need it to deploy and use `NMS`.
I think the base modules should only define the app and its config. Integrations should be handled on a higher level. In fact that's exactly what we're doing with the Juju bundles now.

Reply:
`ingress` relation is moved from  `nms` module to the higher level module.

Reply:
I think both integrations should be removed. The truth is, that the only intended purpose of this module is to be imported by another module. As such, IMO, it should only container the app definition, config (if applicable) and the possible integration endpoints modeled in the `output.tf`. I'm currently testing this idea with COS. Here's the example of the `ingress` configuration for the Alertmanager:
```
resource "juju_integration" "alertmanager-ingress" {
  model = var.model_name

  application {
    name     = module.alertmanager.app_name
    endpoint = module.alertmanager.ingress-endpoint
  }

  application {
    name     = module.traefik.app_name
    endpoint = module.traefik.ingress-endpoint
  }
}
```
WDYT?

Reply:
Removing `ingress` to the higher level is fine. However if the relation puts the charm in `Blocked` status, IMO, it should stay with the charm TF module. In that context, `sdcore-manament` relation could stay with the basic module definition. As a similar approach, DB relation could stay with the NRF app module definition..etc.

``` 
    def _configure_sdcore_nms(self, event: EventBase) -> None:
        """Add Pebble layer and manages Juju unit status.

        Args:
            event (EventBase): Juju event.
        """
        if not self._container.can_connect():
            self.unit.status = WaitingStatus("Waiting for container to be ready")
            return
        if not self.model.relations.get(SDCORE_MANAGEMENT_RELATION_NAME):
            self.unit.status = BlockedStatus(
                f"Waiting for `{SDCORE_MANAGEMENT_RELATION_NAME}` relation to be created"
            )
            return
        if not self._sdcore_management.management_url:
            self.unit.status = WaitingStatus("Waiting for webui management url to be available")
            return
        self._configure_upf_information()
        self._configure_gnb_information()
        self._configure_pebble()
        self.unit.status = ActiveStatus()
````

Reply:
What would be the use of such integration if there's no required application on the other end anyway? What I mean is that if you, for some reason, deploy NMS module alone, it won't work. You still need some higher level module which will put this all together.
I know that what I wrote above would still work if we keep the integration definition together with the module, but imagine a complicated bundle with 20 apps. Instead of just importing required modules and creating integrations between them, you'll spend half of the day trying to figure out which relation are already covered by the individual modules and which you need to create in your module.

Reply:
The integrations are removed.